### PR TITLE
apply observation transformation to replayed exps

### PR DESCRIPTION
**Pull request by hnyu**
I unified algorithm_step for on-policy and off-policy with a little overloading experience with time_step

### CODE REVIEW

File: alf/drivers/policy_driver.py [hunk 1]
```diff
@@ -113,12 +113,20 @@ def get_initial_time_step(self):
...
-    def algorithm_step(self, time_step, state):
+    def algorithm_step(self, time_step, state, training=self._training):
         if self._observation_transformer is not None:
             time_step = time_step._replace(
                 observation=self._observation_transformer(time_step.
                                                           observation))
-        return self._algorithm_step(time_step, state)
+        return self._algorithm_step(time_step, state, training)
+
+    def _algorithm_step(self, time_step, state, training):
```
Comment (emailweixu) [inline alf/drivers/policy_driver.py]: if _algorithm_step() is only used by algorithm_step(), we should merge them into one single function.

File: alf/drivers/policy_driver.py [hunk 2]
```diff
@@ -113,12 +113,20 @@ def get_initial_time_step(self):
                                                      self.env.batch_size)
         return make_action_time_step(time_step, action)
 
-    def algorithm_step(self, time_step, state):
+    def algorithm_step(self, time_step, state, training=self._training):
```
Comment (emailweixu) [inline alf/drivers/policy_driver.py]: Perhaps we should remove the default value of training to make it a little easier to understand the logic.

File: alf/drivers/off_policy_driver.py [hunk 3]
```diff
@@ -47,14 +47,15 @@ def __init__(self,
                 updated after every step in the environment. Each observer is a
                 callable(time_step.Trajectory).
             metrics (list[TFStepMetric]): An optiotional list of metrics.
-            training (bool): True for training, false for evaluating
+            training (bool): by default False because training info is obtained from
```
Comment (emailweixu) [inline alf/drivers/off_policy_driver.py]: Need to comment the meaning of `training`

File: alf/drivers/policy_driver.py (hunk 1 above)
Comment (hnyu) [reply alf/drivers/policy_driver.py]: merged

File: alf/drivers/policy_driver.py (hunk 2 above)
Comment (hnyu) [reply alf/drivers/policy_driver.py]: removed

File: alf/drivers/off_policy_driver.py (hunk 3 above)
Comment (hnyu) [reply alf/drivers/off_policy_driver.py]: removed training flag and set it to False

**Comment by witwolf** [comment]
keyword argument 'training' should be removed here
https://github.com/HorizonRobotics/alf/blob/6ac650530ac9317f7e8d99a32afd1c53750dc97a/alf/drivers/off_policy_driver_test.py#L137-L138

**Comment by hnyu** [comment]
> keyword argument 'training' should be removed here
> https://github.com/HorizonRobotics/alf/blob/6ac650530ac9317f7e8d99a32afd1c53750dc97a/alf/drivers/off_policy_driver_test.py#L137-L138

Remvoed.

**Comment by emailweixu** [comment]
Need to resolve conflict

**Pull request merged**